[DoctrineBridge][DBAL] Add PdoHandler-like $options

hipio opened the pull request:
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | ~ |
| License | MIT |
| Doc PR | https://github.com/symfony/symfony-docs/pull/7132 |

Small PR to incorporate the options that are available to `PdoSessionHandler` into the `DbalSessionHandler`. I do not like being forced to use pre-defined values, especially when they go against the code standards of my current project.
- [x] Update unit tests
- [ ] Gather feedback for my changes

Hipio committed:
[DoctrineBridge] Add PdoHandler-like $options

Small PR to incorporate the options that are available to `PdoSessionHandler` into the `DbalSessionHandler`. I do not like being forced to use pre-defined values, especially when they go against the code standards of my current project.

Hipio committed:
[DoctrineBridge] Adhere to coding standard

fabpot commented on src/Symfony/Bridge/Doctrine/HttpFoundation/DbalSessionHandler.php:
```diff
@@ -56,13 +56,23 @@ class DbalSessionHandler implements \SessionHandlerInterface
         $this->con = $con;
         $this->table = $tableName;
+
+        $this->idCol = isset($options['db_id_col']) ? $options['db_id_col'] : $this->idCol;
```
What about removing the default value of the property and use it here instead, that would be clearer I think.

hipio replied:
That sounds like a good idea. I will update the PR with those changes and be sure to include unit tests within that same commit.

Commit:
Set default values in constructor.

Commit:
Added a test

hipio commented:
I moved the default values to the constructor for clarity, added a test and written documentation for the `DbalSessionHandler` (see Doc PR) as there was none, at least none that was specifically about this class (even though one exists for the `PdoSessionHandler`).

Commit:
Adhere to coding standard

fabpot commented:
Honest question: why would someone use DbalSessionHandler bs PdoSessionHandler?

* This one is slower
* Everything is abstracted, so this one is not more "convenient"
* It is more "complex" to configure as you should ensure that it uses a different connection than the one used by the ORM

hipio commented:
That is very good question!

What I know, which isn't a lot, is that this is used purely for the Doctrine ORM. This means that you do not need to mess with the database to add a session table and everything will be handled through Doctrine itself. 

> It is more "complex" to configure as you should ensure that it uses a different connection than the one used by the ORM

Could you explain this statement? As far as I know I'm using the same connection.

Either way, even if this is worse, the discussion to keep it or not (which is what you seem to be implying) doesn't belong here nor does it affect this PR. Making something more configurable is a good thing, right?

nicolas-grekas commented:
Of side note: the cache component has a PdoAdapter that looks a lot like the PdoSessionHandler; but it also works with DBAL connections (there is no DbalAdapter).

stof commented:
@hipio the first point is a justification for the Schema integration, but this does not justify the handler.

The benefit of the DBAL handler is support for non-PDO drivers of DBAL. But this is the only benefit IMO, and the cache adapter showed that the support could be added in PDOHandler directly (as the DBAL interface matches closely the PDO one)

> Could you explain this statement? As far as I know I'm using the same connection.

if you use the same connection, you will face issues, as the session handlers relies on using transactions.
It would be even totally impossible if the DBAL handler was based on the new PDO handler instead of the legacy one.

hipio commented:
All I want to do is pass options to the DbalSessionHandler because I do not want to be use the pre-defined values. 

If this PR does not contain the right solution to do this, I would like to know what a better solution would be.

fabpot commented:
@hipio This PR does the job :) Sorry for having highjacked your PR with a side discussion that it indeed totally orthogonal to the changes you propose here.

We are in code freeze right now, so this PR will be merged after the release of 3.2.

Tobion commented:
This was already proposed before, see arguments in #13136

fabpot commented:
The main arguments against this change are there: symfony/symfony#14710#issuecomment-104189679

Closing then.

Potherca commented:
@Tobion states:

> This was already proposed before, see arguments in #13136

But I'm not entirely sure I understand what is being said in that thread. As @hipio said:

> If this PR does not contain the right solution to do this, I would like to know what a better solution would be.

I would also like to know, if this is not the way to go, what _is_ the way to implement a fix that will be accepted for configuring the DbalSessionHandler?

stof commented:
@Potherca we started a discussion about the fact that DbalSessionHandler itself is not the way to go. Switch to the PdoSessionHandler, and you will win the configurability.

Potherca commented:
The application I am working with uses Doctrine for everything, this includes sessions. I looks like switching to the PdoSessionHandler will gain me configurability at the cost of loosing the functionality of having session database handling being done by Doctrine.

Or am I missing something?

Potherca commented:
Also, does my comment belong here or should I move it over to #20501?
